odoo/addons/base/models/ir_mail_server.py

```diff
@@ -533,6 +578,94 @@ def send_email(self, message, mail_server_id=None, smtp_server=None, smtp_port=N
+        if not old_email:
+            return new_email
+        old_email = old_email.replace('<', '(').replace('>', ')').replace('"', '')
+        return f'"{old_email}" <{new_email}>'
+
+    @api.model
+    def _find_mail_server(self, mail_from):
+        """Find the appropriate mail server for the given email address.
+
+        Returns: Record<ir.mail_server>, mail_from
+        - Mail server to use to send the email (None if we use the odoo-bin arguments)
+        - Email FROM to use to send the email (in some case, it might be impossible
+          to use the given email address directly if no mail server is configured for)
+        """
+        normalized_mail_from = email_normalize(mail_from)
+        mail_from_domain = email_domain(normalized_mail_from)
+        notifications_email = email_normalize(self._get_default_from_address())
+        notifications_domain = email_domain(notifications_email)
+        notifications_email_full = self._encapsulate_email(mail_from, notifications_email)
+
+        mail_servers = self.sudo().search([], order='sequence')
+
+        # 1. Try to find a mail server for the right mail from
+        mail_server = mail_servers.filtered(lambda m: email_normalize(m.from_filter) == normalized_mail_from)
+        if mail_server:
+            return mail_server[0], mail_from
+
+        mail_server = mail_servers.filtered(lambda m: email_domain_normalize(m.from_filter) == mail_from_domain)
+        if mail_server:
+            return mail_server[0], mail_from
+
+        # 2. Try to find a mail server for <notifications@domain.com>
+        if notifications_email:
+            mail_server = mail_servers.filtered(lambda m: email_normalize(m.from_filter) == notifications_email)
+            if mail_server:
+                return mail_server[0], notifications_email_full
+
+            mail_server = mail_servers.filtered(lambda m: email_domain_normalize(m.from_filter) == notifications_domain)
+            if mail_server:
+                return mail_server[0], notifications_email_full
+
+        # 3. Take the first mail server without "from_filter" because
+        # nothing else has been found... Will spoof the FROM because
+        # we have no other choices
+        mail_server = mail_servers.filtered(lambda m: not m.from_filter)
+        if mail_server:
+            return mail_server[0], mail_from
+
+        # 4. Return the first mail server even if it was configured for another domain
+        if mail_servers:
+            return mail_servers[0], mail_from
+
+        # 5: SMTP config in odoo-bin arguments
+        from_filter = tools.config.get('from_filter')
+
+        if self._match_from_filter(mail_from, from_filter):
+            return None, mail_from
+
+        if notifications_email and self._match_from_filter(notifications_email, from_filter):
+            return None, notifications_email_full
+
+        return None, mail_from
+
+    @api.model
+    def _match_from_filter(self, mail_from, from_filter):
```

Comment:
Maybe we could move it in tools (def email_match_filter) ? That way we have all "tool email functions" available as standalone, we don't really need a self or it to be model-based.

Reply:
Hum, not sure about that, it's very specific to `ir.mail_server`, and will never be used in other models